File: openai_compatible_examples/advanced_usage/openai_sdk_logit_bias.py
"""
Example of using the OpenAI-compatible Chat Completions endpoint with 'logit_bias'
parameter using the official 'openai' Python SDK.

Demonstrates how to influence the likelihood of specific tokens appearing in the response.
"""

import os
import json
from dotenv import load_dotenv
from openai import OpenAI, APIError, APITimeoutError, RateLimitError

# --- Configuration ---
load_dotenv() # Load environment variables from .env file

# Get endpoint and API key from environment variables
API_BASE_URL = os.getenv("OPENAI_API_BASE", "http://localhost:8000/v1")
API_KEY = os.getenv("OPENAI_API_KEY", "dummy-key")
MODEL_NAME = os.getenv("MODEL_NAME") # Optional: If endpoint supports model selection

# --- Initialize OpenAI Client ---
# Point the client to the custom endpoint
client = OpenAI(
    base_url=API_BASE_URL,
    api_key=API_KEY,
)

# --- Logit Bias Setup ---
# Logit bias maps token IDs (integers) to bias values (floats from -100 to 100).
# Positive values increase likelihood, negative values decrease likelihood.
# A bias of 100 effectively guarantees the token, -100 effectively bans it.

# IMPORTANT: You need the actual integer token IDs used by the specific model
#            you are targeting. These IDs vary between tokenizers (e.g., GPT-4,
#            Claude, Llama). Finding these IDs usually requires using the
#            tokenizer associated with the model (like tiktoken for OpenAI models).
#
# Example (Illustrative - Replace with actual token IDs for your model):
# Let's assume we found the token IDs for " red" and " blue" (note potential leading space):
#   " red": 1234
#   " blue": 5678
# Let's also assume the token ID for " green" is 9101

# We want to make " red" more likely and ban " green"
# The SDK expects a dictionary where keys are integer token IDs.
example_logit_bias = {
    # 1234: 5,    # Increase likelihood of " red" (Use actual token ID)
    # 9101: -100  # Ban " green" (Use actual token ID)
    # --- Replace above with actual token IDs for your model --- 
    # Placeholder for demonstration - this won't work without correct IDs
    1598: 10,    # Example: Make token ID 1598 more likely 
    8481: -100,  # Example: Ban token ID 8481
}

# --- API Request ---
messages = [
    {"role": "user", "content": "What is your favorite color? Describe it."}
]

print("--- Sending request to model with Logit Bias ---")
print(f"Messages: {messages}")
print(f"Logit Bias (using placeholder token IDs): {example_logit_bias}")
print("\nNOTE: Logit bias requires using the correct integer TOKEN IDs for the target model.")
print("The example bias values are placeholders and may not affect the output correctly.")
print("You must replace the keys in 'logit_bias' with actual token IDs.")
print("-" * 30)

try:
    completion = client.chat.completions.create(
        model=MODEL_NAME, # Required for SDK
        messages=messages,
        logit_bias=example_logit_bias,
        max_tokens=50,
        temperature=0.7,
    )

    print("--- Full API Response ---")
    # Use model_dump_json for cleaner output of Pydantic models
    print(completion.model_dump_json(indent=2))
    print("-" * 30)

    # --- Response Handling ---
    assistant_message = completion.choices[0].message.content
    print(f"Assistant Message (with bias applied):")
    print(assistant_message)

    # You would observe if the output favors tokens with positive bias
    # and avoids tokens with negative bias (especially -100).

except (APIError, RateLimitError, APITimeoutError) as e:
    print(f"An API error occurred: {e}")
    if hasattr(e, 'status_code'):
        print(f"Status Code: {e.status_code}")
    if hasattr(e, 'response') and e.response is not None:
        try:
            print(f"Response Body: {e.response.text}")
        except Exception:
             print("Could not print response body.")
    elif hasattr(e, 'message'):
        print(f"Error Message: {e.message}")

except KeyError as e:
    print(f"Error accessing expected key in API response: {e}")
    print("Response structure might be different than expected.")
    print(completion.model_dump_json(indent=2) if 'completion' in locals() else "No response object")

except Exception as e:
    print(f"An unexpected error occurred: {e}")
    print(f"Type: {type(e)}")

print("-" * 30)
print("Logit bias example complete.") 

File: java_sdk_examples/src/main/java/com/example/advanced_usage/OpenAiSdkLogitBias.java
package com.example.advanced_usage;

import com.example.utils.ConfigLoader;
import com.theokanning.openai.completion.chat.ChatCompletionRequest;
import com.theokanning.openai.completion.chat.ChatCompletionChoice;
import com.theokanning.openai.completion.chat.ChatMessage;
import com.theokanning.openai.completion.chat.ChatMessageRole;
import com.theokanning.openai.service.OpenAiService;

import java.io.IOException;
import java.time.Duration;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.theokanning.openai.client.OpenAiApi;
import static com.theokanning.openai.service.OpenAiService.*;

public class OpenAiSdkLogitBias {

    public static void main(String[] args) {
        String apiKey = ConfigLoader.getApiKey();
        String apiBaseUrl = ConfigLoader.getApiBaseUrl();
        String modelName = ConfigLoader.getModelName();

        if ("DUMMY_API_KEY".equals(apiKey)) {
            System.err.println("API Key not configured. Please check your .env file.");
            return;
        }

        System.out.println("--- TheoKanning/openai-java SDK Logit Bias Example --- ");

        OpenAiService service;
        try {
            if (apiBaseUrl != null && !apiBaseUrl.equals("https://api.openai.com/v1/") && !apiBaseUrl.isEmpty()) {
                ObjectMapper mapper = defaultObjectMapper();
                OkHttpClient client = defaultClient(apiKey, Duration.ofSeconds(30)).newBuilder().build();
                Retrofit retrofit = defaultRetrofit(client, mapper).newBuilder().baseUrl(apiBaseUrl).build();
                OpenAiApi openAiApi = retrofit.create(OpenAiApi.class);
                service = new OpenAiService(openAiApi);
            } else {
                service = new OpenAiService(apiKey, Duration.ofSeconds(30));
            }
        } catch (Exception e) {
            System.err.println("Error initializing OpenAiService: " + e.getMessage());
            e.printStackTrace();
            return;
        }

        List<ChatMessage> messages = new ArrayList<>();
        messages.add(new ChatMessage(ChatMessageRole.USER.value(), "What is the color of the sky on a sunny day? Try to avoid common answers."));

        // Logit bias: Increase probability of token ID for "azure" (e.g., 26583 - this ID is an EXAMPLE and will vary by model/tokenizer)
        // Decrease probability of token ID for "blue" (e.g., 5012 - EXAMPLE ID)
        // Values are from -100 to 100. Higher values make token more likely, lower values make it less likely.
        // IMPORTANT: Token IDs are model-specific. You need to find the correct token IDs for your model.
        // This is a conceptual example. Finding exact token IDs usually requires a tokenizer for the specific model.
        Map<String, Integer> logitBias = new HashMap<>();
        logitBias.put("26583", 50);  // Hypothetical token ID for "azure", strongly increase
        logitBias.put("azure", 50); // Some SDKs might allow string keys if they handle tokenization.
                                  // TheoKanning SDK 0.20.0 expects String token IDs in the map.
        logitBias.put("5012", -50); // Hypothetical token ID for "blue", strongly decrease
        logitBias.put("blue", -50); // Again, for concept.

        System.out.println("Note: Logit bias token IDs are model-specific. The IDs used here (26583, 5012) are illustrative examples.");
        System.out.println("Effective logit bias will depend on correct tokenization for the target model and SDK handling.");

        ChatCompletionRequest.Builder requestBuilder = ChatCompletionRequest.builder()
                .messages(messages)
                .temperature(0.7)
                .logitBias(logitBias) // Set the logit bias
                .n(1);

        if (modelName != null && !modelName.isEmpty()) {
            requestBuilder.model(modelName);
        } else {
            System.out.println("No model specified, using 'gpt-3.5-turbo'.");
            requestBuilder.model("gpt-3.5-turbo");
        }
        ChatCompletionRequest request = requestBuilder.build();

        System.out.println("Sending request with logit bias...");
        System.out.println("Model: " + request.getModel());
        System.out.println("Logit Bias Applied (Conceptual): " + logitBias);
        System.out.println("------------------------------------");

        try {
            List<ChatCompletionChoice> choices = service.createChatCompletion(request).getChoices();
            System.out.println("--- API Response --- ");
            if (choices != null && !choices.isEmpty()) {
                System.out.println("Assistant Message: " + choices.get(0).getMessage().getContent());
            } else {
                 System.out.println("No choices returned.");
            }
            System.out.println("------------------------------------");

        } catch (retrofit2.HttpException e) {
            System.err.println("API HTTP Error: " + e.code() + " " + e.message());
             try {
                System.err.println("Error Body: " + (e.response() != null && e.response().errorBody() != null ? e.response().errorBody().string() : "N/A"));
            } catch (IOException ioe) { System.err.println("Error reading error body: " + ioe); }
            e.printStackTrace();
        } catch (Exception e) {
            System.err.println("An unexpected error occurred: " + e.getMessage());
            e.printStackTrace();
        }

        service.shutdownExecutor();
        System.out.println("SDK Logit Bias example complete.");
    }
} 